### PR TITLE
Enable end-to-end tests for pull requests

### PR DESCRIPTION
**Pull request by andresmgot**
This PR adds a new job that run our end-to-end tests (helm tests) that's run in a local Kubernetes cluster using [kind](https://github.com/kubernetes-sigs/kind). This enable us to use our end-to-end tests with untrusted code (PRs).

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -117,17 +124,37 @@ build_images: &build_images
...
     - <<: *exports
     - run: |
+        mkdir -p images/
         read -ra IMG_ARRAY <<< "$IMAGES"
         if [[ -n "${CIRCLE_TAG}" ]]; then
           makeArgs="VERSION=${CIRCLE_TAG}"
         fi
         for IMAGE in "${IMG_ARRAY[@]}"; do
-          make IMG_MODIFIER="$IMG_MODIFIER" IMAGE_TAG="${DEV_TAG}" $makeArgs ${IMAGE}
+          make IMG_MODIFIER="$IMG_MODIFIER" IMAGE_TAG="${DEV_TAG}" $makeArgs kubeapps/${IMAGE}
```
Comment (andresmgot) [inline .circleci/config.yml]: Note: I have removed `kubeapps/` from `${IMAGE}` to be able store `${IMAGE}` as a file not `folder/file`. Let me know if that is readable or you prefer to keep using the previous definition of `${IMAGE}`

File: script/e2e-test.sh [hunk 2]
```diff
@@ -31,7 +31,10 @@ echo "IMAGE_REPO_SUFFIX: $IMG_MODIFIER"
 kubectl version
 
 # Install Tiller with TLS support
+kubectl -n kube-system create sa tiller
```
Comment (andresmgot) [inline script/e2e-test.sh]: we need to set the tiller service account as cluster-admin because by default, in a cluster created with `kind` the `default` service account in the `kube-system` namespace doesn't have permissions to read config maps in that namespace.

File: .circleci/config.yml (hunk 1 above)
Comment (prydonius) [reply .circleci/config.yml]: The only thing that might be an issue is if we want to push these images somewhere other than Docker Hub in the future, or need to use different repos for different images, but I don't think that's the case right now.

File: .circleci/config.yml [hunk 3]
```diff
@@ -117,17 +124,37 @@ build_images: &build_images
...
           if [[ -z "${CIRCLE_PULL_REQUEST}" && -n "${DOCKER_USERNAME}" && -n "${DOCKER_PASSWORD}" ]]; then
             docker login -u="${DOCKER_USERNAME}" -p="${DOCKER_PASSWORD}"
-            docker push ${IMAGE}${IMG_MODIFIER}:${DEV_TAG}
+            docker push kubeapps/${IMAGE}${IMG_MODIFIER}:${DEV_TAG}
           fi
+          docker save kubeapps/${IMAGE}${IMG_MODIFIER}:${DEV_TAG} > images/${IMAGE}${IMG_MODIFIER}:${DEV_TAG}
```
Comment (prydonius) [inline .circleci/config.yml]: We only need to do this if this is a pull request, right?

File: .circleci/config.yml [hunk 4]
```diff
@@ -217,20 +232,44 @@ jobs:
...
+      IMAGES: "dashboard"
     <<: *build_images
   release:
     docker:
       - image: circleci/golang:1.11
     steps:
       - checkout
       - run: REPO_DOMAIN=kubeapps REPO_NAME=kubeapps ./script/create_release.sh ${CIRCLE_TAG}
+  local_e2e_tests:
+    working_directory: /home/circleci/.go_workspace/src/github.com/kubeapps/kubeapps
```
Comment (prydonius) [inline .circleci/config.yml]: why is this necessary?

File: .circleci/config.yml [hunk 5]
```diff
@@ -217,20 +232,44 @@ jobs:
...
+    steps:
+      - checkout
+      - <<: *exports
+      # By default Go 1.9 is installed, we need 1.11 to use kind
+      - run: |
+          sudo rm -rf /usr/local/go/*
+          sudo chown $(whoami) /usr/local/go
+          curl -LO https://dl.google.com/go/go1.11.9.linux-amd64.tar.gz
+          tar -C /usr/local -xzf go1.11.9.linux-amd64.tar.gz
+      - run: go get -u sigs.k8s.io/kind
```
Comment (prydonius) [inline .circleci/config.yml]: Can we use a pre-built binary from a release instead? https://github.com/kubernetes-sigs/kind/releases. This should avoid the need to install Go 1.11?

File: script/e2e-test.sh (hunk 2 above)
Comment (prydonius) [reply script/e2e-test.sh]: Weird, I thought this was the case in GKE too - not sure how we were able to avoid it before?

File: .circleci/config.yml (hunk 1 above)
Comment (andresmgot) [reply .circleci/config.yml]: Yes, we can change this later if needed.

File: .circleci/config.yml (hunk 3 above)
Comment (andresmgot) [reply .circleci/config.yml]: It's mandatory when it is a PR. When running this in master we could pull the images but this is convenient because loading the images from the local file is faster.

File: .circleci/config.yml (hunk 4 above)
Comment (andresmgot) [reply .circleci/config.yml]: it's probably not, we are not compiling in this job

File: .circleci/config.yml (hunk 5 above)
Comment (andresmgot) [reply .circleci/config.yml]: the only release available is a prerelease. I am not sure if it will work but I can try, probably it's enough.

File: script/e2e-test.sh (hunk 2 above)
Comment (andresmgot) [reply script/e2e-test.sh]: tbh I am not sure. My supposition is that depending on the user that creates the cluster, GKE default permissions fallback to those, so if an user with privileges creates a cluster (like the account set here) it's not necessary to create the extra service account. I have tried with my personal account and for me it requires to use the tiller serviceaccount.

Comment (prydonius) [reply script/e2e-test.sh]: Hmm that's pretty strange, so the service account here has admin privileges in the project?

File: .circleci/config.yml (hunk 3 above)
Comment (prydonius) [reply .circleci/config.yml]: Ah I was thinking we only do the local test for PRs, but we actually run it all the time. sg.

**Comment by prydonius** [comment]
Nice, looks like the prebuilt version works. I believe they're marking all 0.x releases as pre-releases in GitHub.

File: script/e2e-test.sh (hunk 2 above)
Comment (andresmgot) [reply script/e2e-test.sh]: yes, it has more privileges than a normal account
